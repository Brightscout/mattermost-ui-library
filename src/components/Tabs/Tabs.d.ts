import {TabsProps as BootstrapTabProps, TabProps} from 'react-bootstrap';

export interface TabsConfig {

    /**
   * Key to determine each tab uniquely
   */
    key: TabProps['eventKey'];

    /**
     * Title to be displayed on tab button
     */
    title: TabProps['title'];

    /**
     * Content to be displayed when tab is opened
     */
    content: TabProps['children'];
}

export interface TabsProps extends BootstrapTabProps {

    /**
     * Configuration for rendering tabs
     */
    tabsConfig: TabsConfig[];

    /**
     * Have all tabs proportionately fill all available widths.
     * @default false
     */
    fill?: boolean;

    /**
     * Have all tabs evenly fill all available widths
     * @default true
     */
    justify?: boolean;

    /**
     * Flag to make tabs header sticky
     * @default true
     */
    stickyHeader?: boolean;
}
