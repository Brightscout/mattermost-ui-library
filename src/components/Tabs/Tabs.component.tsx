import React from 'react';
import Tab from 'react-bootstrap/Tab';

import {StyledTabs} from './Tabs.styles';
import {TabsProps} from './Tabs';

import 'bootstrap/dist/css/bootstrap.min.css';

export const Tabs = (props: TabsProps) => {
    const {
        tabsConfig,
        fill = false,
        justify = true,
        mountOnEnter = true,
        unmountOnExit = true,
        defaultActiveKey,
        stickyHeader = true,
        ...restProps
    } = props;

    return (
        <StyledTabs
            stickyHeader={stickyHeader}
            fill={fill}
            justify={justify}
            mountOnEnter={mountOnEnter}
            unmountOnExit={unmountOnExit}
            defaultActiveKey={defaultActiveKey}
            {...restProps}
        >
            {tabsConfig.map((tab) => (
                <Tab
                    key={tab.key}
                    eventKey={tab.key}
                    title={tab.title}
                >
                    {tab.content}
                </Tab>))}
        </StyledTabs>
    );
};

