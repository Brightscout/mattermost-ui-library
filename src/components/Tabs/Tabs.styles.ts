import styled from 'styled-components';
import {TabProps as BootstrapTabProps, Tabs} from 'react-bootstrap';

import colors from '@Styles/colorsForJs.module.scss';

export const StyledTabs = styled(Tabs).withConfig({
    shouldForwardProp: (prop) => prop !== 'stickyHeader' as keyof BootstrapTabProps,
})<{stickyHeader: boolean}>(({stickyHeader}) => ({

    '.nav-tabs.nav-justified > li > a': {
        marginBottom: 0,
        borderColor: 'transparent',
    },

    '& .nav-tabs': {
        borderBottom: `2px solid ${colors.tabsHeaderBorder}`,
        marginBottom: '24px',

        ...(stickyHeader && {
            position: stickyHeader ? 'sticky' : 'initial',
            top: 0,
            background: colors.centerChannelBg,
        }),
    },

    '& .nav-item': {
        marginBottom: '-2px',
        lineHeight: '20px',
        fontWeight: '500',
        backgroundColor: 'transparent',

        a: {
            color: colors.centerChannel,

            ':hover, :focus': {
                color: colors.primary,
                backgroundColor: 'transparent',
                borderColor: 'transparent',
            },
        },

        '&.active': {
            borderBottom: `2px solid ${colors.primary}`,

            a: {
                color: colors.centerChannel,
                borderColor: 'transparent',

                ':hover, :focus': {
                    backgroundColor: 'transparent',
                    borderColor: 'transparent',
                },
            },
        },
    },
}));
