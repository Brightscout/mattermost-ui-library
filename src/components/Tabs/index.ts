export {Tabs} from './Tabs.component';
export type {TabsProps} from './Tabs';
