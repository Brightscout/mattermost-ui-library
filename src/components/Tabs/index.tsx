import React from 'react';
import {Tabs as ReactBootstrapTabs, Tab} from 'react-bootstrap';

type TabsProps = {
    tabs: TabData[];
    tabsClassName?: string;
}

const Tabs = ({
    tabs,
    tabsClassName = '',
}: TabsProps) => (
    <ReactBootstrapTabs className={`tabs ${tabsClassName}`}>
        {tabs.map((tabData, index) => (
            <Tab
                key={tabData.title}
                eventKey={index}
                title={tabData.title}
            >
                {tabData.tabPanel}
            </Tab>
        ))}
    </ReactBootstrapTabs>
);

export default Tabs;
