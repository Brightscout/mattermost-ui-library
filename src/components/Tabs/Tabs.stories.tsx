import React from 'react';
import {Story, Meta} from '@storybook/react';

import {Tabs} from './Tabs.component';
import {TabsConfig, TabsProps} from './Tabs';

export default {
    title: 'Component/Tabs',
    component: Tabs,
} as Meta<TabsProps>;

const TabsTemplate: Story<TabsProps> = (args) => <Tabs {...args}/>;

export const Default = TabsTemplate.bind({});

export const WithoutStickyHeader = TabsTemplate.bind({});

export const WithJustified = TabsTemplate.bind({});

const tabsConfig: TabsConfig[] = [
    {
        key: 1,
        title: 'Tab 1',
        content: <p style={{height: '150vh'}}>{'This is content for Tab 1'}</p>,
    },
    {
        key: 2,
        title: 'Tab 2',
        content: <p style={{height: '150vh'}}>{'This is content for Tab 2'}</p>,
    },
];

Default.args = {
    tabsConfig,
};

WithJustified.args = {
    ...Default.args,
    justified: true,
};

WithoutStickyHeader.args = {
    ...Default.args,
    stickyHeader: false,
};
