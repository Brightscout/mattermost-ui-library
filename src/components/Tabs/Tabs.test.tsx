import React from 'react';
import {shallow, ShallowWrapper} from 'enzyme';

import Tabs from '.';

// Mock props data
const tabPanelClass = 'tab-panel-class';
const tabsProps = {
    tabs: [{
        title: 'Tab 1',
        tabPanel: <div className={tabPanelClass}>{'tab-panel-1'}</div>,
    }, {
        title: 'Tab 2',
        tabPanel: <div className={tabPanelClass}>{'tab-panel-2'}</div>,
    }],
    tabsClassName: 'tabs-class',
};

describe('ToggleSwitch', () => {
    let component: ShallowWrapper<any, Readonly<{}>, React.Component<{}, {}, any>>;

    beforeEach(() => {
        component = shallow(<Tabs {...tabsProps} />);
    });

    it('Should render correctly', () => {
        expect(component).toMatchSnapshot();
    });

    it('Should apply the passed className prop', () => {
        expect(component.hasClass(tabsProps.tabsClassName)).toBeTruthy();
    });

    it('Should render all the passed tabs', () => {
        expect(component.find(`.${tabPanelClass}`)).toHaveLength(tabsProps.tabs.length);
    });
});
