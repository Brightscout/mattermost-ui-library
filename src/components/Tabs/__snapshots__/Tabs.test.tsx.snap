// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tabs Should render correctly 1`] = `
<Tabs
  className="tabs tabs-class"
  mountOnEnter={false}
  unmountOnExit={false}
  variant="tabs"
>
  <Tab
    eventKey={0}
    key="Tab 1"
    title="Tab 1"
  >
    <div
      className="tab-panel-class"
    >
      tab-panel-1
    </div>
  </Tab>
  <Tab
    eventKey={1}
    key="Tab 2"
    title="Tab 2"
  >
    <div
      className="tab-panel-class"
    >
      tab-panel-2
    </div>
  </Tab>
</Tabs>
`;
